paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/CdcRecordStoreMultiWriteOperator.java

```diff
@@ -0,0 +1,176 @@
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.apache.paimon.flink.sink.cdc.CdcRecordStoreWriteOperator.RETRY_SLEEP_TIME;
+
+/**
+ * A {@link PrepareCommitOperator} to write {@link CdcRecord}. Record schema may change. If current
+ * known schema does not fit record schema, this operator will wait for schema changes.
+ */
+public class CdcRecordStoreMultiWriteOperator extends PrepareCommitOperator<MultiplexCdcRecord> {
+
+    private static final long serialVersionUID = 1L;
+    private final Catalog catalog;
+    private final StoreSinkWrite.Provider storeSinkWriteProvider;
+    private final String initialCommitUser;
+    private FileStoreTable table;
+    private Map<Identifier, FileStoreTable> tables;
+    private StoreSinkWriteState state;
+    private StoreSinkWrite write;
+    private Map<Identifier, StoreSinkWrite> writes;
+    private String commitUser;
+
+    public CdcRecordStoreMultiWriteOperator(
+            Catalog catalog,
+            StoreSinkWrite.Provider storeSinkWriteProvider,
+            String initialCommitUser) {
+        this.catalog = catalog;
+        this.storeSinkWriteProvider = storeSinkWriteProvider;
+        this.initialCommitUser = initialCommitUser;
+    }
+
+    @Override
+    public void initializeState(StateInitializationContext context) throws Exception {
+        super.initializeState(context);
+
+        // Each job can only have one user name and this name must be consistent across restarts.
+        // We cannot use job id as commit user name here because user may change job id by creating
+        // a savepoint, stop the job and then resume from savepoint.
+        commitUser =
+                StateUtils.getSingleValueFromState(
+                        context, "commit_user_state", String.class, initialCommitUser);
+
+        state = new StoreSinkWriteState(context, (tableName, partition, bucket) -> true);
+        this.tables = new HashMap<>();
+        this.writes = new HashMap<>();
+    }
+
+    @Override
+    public void processElement(StreamRecord<MultiplexCdcRecord> element) throws Exception {
+        MultiplexCdcRecord record = element.getValue();
+
+        String databaseName = record.getDatabaseName();
+        String tableName = record.getTableName();
+        Identifier key = Identifier.create(databaseName, tableName);
+        table =
+                tables.computeIfAbsent(
+                        key,
+                        id -> {
+                            try {
+                                return (FileStoreTable) catalog.getTable(id);
+                            } catch (Catalog.TableNotExistException e) {
+                                return null;
+                            }
+                        });
+
+        if (table == null) {
```

Comment:
here should be a `while (true) {...}` to wait table creation by `UpdatedDataFieldsProcessFunction`.

Reply:
For MySQL CDC:

1. In snapshot phase, any DDL changes (including new table added) will result in Source table schema inconsistency and the job will fail, so we only consider ddl in incremental phase.

2. In incremental phase, it will only use 1 parallelism (source reader) to subscribe binlog, so the process function that create the new table in a blocking manner strictly `happen-before` this code segment.

Reply:
The process function is in a separate task, it is not a blocking manner.

Reply:
That makes sense...

Changed the code and added async table create test.